docs/src/proposals/multi-writers.md

```diff
@@ -0,0 +1,188 @@
+store data directly from the received shred and do not rely on any previous
+metadata.  As a result, they are simply write opreations which do not require
+merge operation to ensure potential read-modify-write atomicity.
+
+### SlotMeta
+The SlotMeta column family involves in intensive read-modify-write operation
+and requires converting to merge operator in order to ensure atomicity without
+locking.  Here're how we can ensure thread-safety for each field:
+
+#### [consumed](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L17-L19):
+This field is [updated](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1460-L1469)
+based on the in-memory BTreeSet `data_index` by checking whether we've consumed
+all shreds up to the current index.  While the consumed field does not require
+additional work for converting from put operation to merge operation, we must
+ensure the thread-safety of `data_index`.
+
+#### [received](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L20-L23)
+received field is updated by taking the max of the current index + 1 and the
+existing value [here](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3180-L3182).
+To correctly handle a potential data race where the later smaller value
+overrides the older bigger value, a merge operator that implements such max
+operation should be used.
+
+#### [first_shred_timestamp](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L24-L25)
+Similar to the received field, `first_shred_timestamp` can use merge operator
+that includes min operation to correctly capture the `first_shred_timestamp` to
+avoid the case where more than one writers are eligible for updating the
+`first_shred_timestamp` field to avoid potential override in case we want to
+keep this field accurate.
+
+#### [last_index](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L26-L27)
+The last_index field doesn't require merge operator to ensure its atomicity as
+there should be only one shred which has its [is_last_in_slot to be true]
+(https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3189-L3200).
+
+#### [parent_slot](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore_meta.rs#L28-L29)
+parent_slot is updated only from [std::u64::MAX to a valid parent slot](https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L3251-L3253).
+A merge operator is needed to ensure we never override a valid parent slot
+by u64::MAX.
```

Comment:
Just an fyi: Note it's possible that a malicious leader could send two shreds for the same slot with a different `parent_slot` field. We don't currently handle that in the checks either 😃 https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1162. 

Right now we take the parent of the first shred we receive for the slot `N`. So if the first shred in slot `N` has the incorrect parent, there will be an error in replay (because the PoH hash verification chain will fail from the wrong parent), and the validator will mark the slot `N` as dead. Other validators that get the correct parent first will successfully replay.

The only way this partition is successfully resolved is if each validator sees the alternate version of the duplicate shred and calls `handle_duplicate` here: https://github.com/solana-labs/solana/blob/b57097ef18ab250aeb4ea8d3d111748ab4c41440/ledger/src/blockstore.rs#L1197